server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java

```diff
@@ -260,6 +261,175 @@ public void testEpochMillisParser() {
             assertThat(formatter.format(instant), is("-0.12345"));
             assertThat(Instant.from(formatter.parse(formatter.format(instant))), is(instant));
         }
+        {
+            Instant instant = Instant.from(formatter.parse("-86400000"));
+            assertThat(instant.getEpochSecond(), is(-86400L));
+            assertThat(instant.getNano(), is(0));
+            assertThat(formatter.format(instant), is("-86400000"));
+            assertThat(Instant.from(formatter.parse(formatter.format(instant))), is(instant));
+        }
+        {
+            Instant instant = Instant.from(formatter.parse("-86400000.999999"));
+            assertThat(instant.getEpochSecond(), is(-86401L));
+            assertThat(instant.getNano(), is(999000001));
+            assertThat(formatter.format(instant), is("-86400000.999999"));
+            assertThat(Instant.from(formatter.parse(formatter.format(instant))), is(instant));
+        }
+
+        // validate that negative epoch millis around rounded appropriately by the parser
+        LongSupplier supplier = () -> 0L;
+        {
+            Instant instant = formatter.toDateMathParser().parse("0", supplier, true, ZoneId.of("UTC"));
+            assertEquals("1970-01-01T00:00:00.000999999Z", instant.toString());
```

Comment:
For this bug I don't think it matters, but to follow the docs and if there is some other edge case with date comparison elsewhere, the date created should follow the spec outlined in the [range query docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#missing-date-components) where when rounded nanos are 999_999. 

Currently the DateFieldMapper compares at a millisecond level and so keeping 999_999 nanos has no impact for this specific case (it can be dropped at this point); although one might argue that this is a separate bug as all (I believe all of them) of our date fields in mapping support nanos and so if a user were to store and compare nanos they would find our resolution of milliseconds buggy as in and I was planning to create a separate issue for this:  

```
curl -XPUT --header 'Content-Type: application/json' "http://localhost:9200/test" -d '{
  "mappings": {
    "properties": {
      "mydate": {
        "type": "date",
        "format": "yyyy-MM-dd HH:mm:ss.SSSSSSSSSX"
      }
    }
  }
}'
```

```
curl -XPUT --header 'Content-Type: application/json' "http://localhost:9200/test/_doc/1?refresh" -d '{
  "mydate": "1969-12-31 00:00:00.000000001Z"
}'
```

both of these queries should produce a result of 1 but instead produce a result of 0 because our resolution for date comparison only goes down to the millisecond level in DateFieldMapper.

```
# 1969-12-31Z as ms since epoch
curl -XGET --header 'Content-Type: application/json' "http://localhost:9200/test/_search?filter_path=hits.total.value" -d '{
  "query": {
    "range": {
      "mydate": {
        "gt": "-86400000.000000",
        "format": "epoch_millis"
      }
    }
  }
}'
```

```
# 1969-12-31Z as ms since epoch
curl -XGET --header 'Content-Type: application/json' "http://localhost:9200/test/_search?filter_path=hits.total.value" -d '{
  "query": {
    "range": {
      "mydate": {
        "gt": "-86400000",
        "format": "epoch_millis"
      }
    }
  }
}'
```

Reply:
@rjernst any additional questions or concerns here?